task/host-sp-comms/src/main.rs

```diff
@@ -250,6 +250,11 @@ struct ServerImpl {
     reboot_state: Option<RebootState>,
     host_kv_storage: HostKeyValueStorage,
     hf_mux_state: Option<HfMuxState>,
+    /// Set to true when the host OS panics, and unset when the system reboots.
+    ///
+    /// This is used to determine whether a host-triggered power-off is due to a
+    /// kernel panic or was a normal power-off.
+    host_just_panicked: bool,
```

Comment:
Yeah, I think that's a good idea. I'll add boot failure as well.

Reply:
They are two different reasons but a host boot failure will be immediately followed by a panic. I still think it's useful to distinguish the two (you'd have to just treat boot failure as more relevant).

Of course a panic may not be immediately followed by a request for a reboot - if the kernel debugger is loaded we'll end up at a debugger prompt until somebody intervenes.

Reply:
We should now understand host boot failure as a separate thing from host panics in https://github.com/oxidecomputer/hubris/pull/1953/commits/2650ced4a3bcfaf0ef1384ee64b1f0961e794e2b. We don't set the panic flag if the boot failure flag was already set, so I think we should be able to differentiate between them correctly.

If there's an easy way to force a boot failure on a bench system, I'd love to test that Hubris is able to understand all of these events correctly?

Reply:
> If there's an easy way to force a boot failure on a bench system

I've done this on accident many times! If you flash a mismatched phase 1 ROM to whatever is in the phase 2 BSU of the host, it'll go into a boot failure loop.

Reply:
Okay, I did some testing on a bench gimlet that has no host phase 2 image, so booting will fail predictably (thanks @rmustacc!), and it looks like the current code is treating the boot failure as a host panic:
```
  44  683        1        1 SetState { prev: A0PlusHP, next: A2, why: HostPanic, now: 0xa3941 }
  45 1422        1        1 V3P3SysA0VOut(Volts(3.3105469))
  46  692        1        1 PGStatus { b_pg: 0xff, c_pg: 0x3, nic: 0x7f }
  47  698        1        1 SMStatus { a1: 0x5, a0: 0xc }
  48  703        1        1 PowerControl(0x3)
  49  707        1        1 InterruptFlags(0x0)
  50  670        1        1 UpdateState(A2)
  51 1422        1        1 V3P3SysA0VOut(Volts(3.3105469))
  52  920        1        1 A2
  53 1422        1        1 V3P3SysA0VOut(Volts(1.1875))
  54 1422        1        1 V3P3SysA0VOut(Volts(0.9765625))
  55  534        1        1 Status { ier: 0x0, ifr: 0x0, amd_status: 0xb, amd_a0: 0xb }
  56  683        1        1 SetState { prev: A2, next: A0, why: HostReboot, now: 0xa4d9d }
```
so I've clearly messed something up in here...

Reply:
hahahaha, whoops: https://github.com/oxidecomputer/hubris/blob/2650ced4a3bcfaf0ef1384ee64b1f0961e794e2b/task/host-sp-comms/src/main.rs#L827-L830

Reply:
There's also a question about whether it's nicer for `host_sp_comms` to continue to use the same `reason` value when it sets the system back to A0 after a host boot failure/panic. Currently, we always set the reason for transitions back to A0 driven by `host_sp_comms` to `HostReboot`, because we clear the last power off reason when Jefe tells us that we've gone to A2. But perhaps we should instead wait to clear that until we get to A0 again?

Reply:
Okay, here we go; *now* it does the right thing (as of 108894718aec404173b363f5c6f77658e3af8dbb), here's an excerpt from the sequencer ringbuffer when a host boot fails:
```
  44  683        1        1 SetState { prev: A0PlusHP, next: A2, why: HostBootFailure, now: 0xb580a }
  45 1422        1        1 V3P3SysA0VOut(Volts(3.3105469))
  46  692        1        1 PGStatus { b_pg: 0xff, c_pg: 0x3, nic: 0x7f }
  47  698        1        1 SMStatus { a1: 0x5, a0: 0xc }
  48  703        1        1 PowerControl(0x3)
  49  707        1        1 InterruptFlags(0x0)
  50  670        1        1 UpdateState(A2)
  51 1422        1        1 V3P3SysA0VOut(Volts(3.2480469))
  52  920        1        1 A2
  53 1422        1        1 V3P3SysA0VOut(Volts(1.1953125))
  54 1422        1        1 V3P3SysA0VOut(Volts(0.9921875))
  55  534        1        1 Status { ier: 0x0, ifr: 0x0, amd_status: 0xb, amd_a0: 0xb }
  56  683        1        1 SetState { prev: A2, next: A0, why: HostBootFailure, now: 0xb6c66 }
```

The whole ringbuffer is also interesting, as you can also see the `SetState` call with `why: Other` when I manually poke the machine into A0 using `humility hiffy`:

<details>
<summary>Full sequencer ringbuffer contents</summary>

```
 NDX LINE      GEN    COUNT PAYLOAD
   0  273        1        1 Ice40Rails(true, true)
   1  302        1        1 Ice40PowerGoodV1P2(true)
   2  323        1        1 Ice40PowerGoodV3P3(true)
   3  365        1        1 IdentValid(false)
   4  368        1        1 ChecksumValid(false)
   5  371        1        1 Reprogram(true)
   6  385        1        1 Programming
   7  414        1        1 Programmed
   8  418        1        1 RailsOff
   9  421        1        1 Ident(0xde01)
  10  427        1        1 A2Status(0x0)
  11  450        1      446 ClockConfigWrite
  12  461        1        1 ClockConfigSuccess
  13 1422        1        1 V3P3SysA0VOut(Volts(0.10546875))
  14  463        1        1 A2
  15 1235        1        1 SpdDimmsFound(0x10)
  16  683        1        1 SetState { prev: A2, next: A0, why: Other, now: 0x865dd }
  17 1422        1        1 V3P3SysA0VOut(Volts(0.091796875))
  18  692        1        1 PGStatus { b_pg: 0x0, c_pg: 0x0, nic: 0x0 }
  19  698        1        1 SMStatus { a1: 0x0, a0: 0x0 }
  20  703        1        1 PowerControl(0x0)
  21  707        1        1 InterruptFlags(0x0)
  22  749        1        1 A1Status(0x2)
  23  749        1      100 A1Status(0x3)
  24  749        1        1 A1Status(0x5)
  25  767        1        1 CPUPresent(true)
  26  777        1        1 Coretype { coretype: true, sp3r1: true, sp3r2: false }
  27  805        1       11 A0Status(0x1)
  28  805        1      284 A0Status(0x2)
  29  805        1        1 A0Status(0x4)
  30  805        1        2 A0Status(0x6)
  31  805        1        1 A0Status(0x7)
  32  827        1        1 RailsOn
  33  839        1        1 A0Power(0x7)
  34  839        1        9 A0Power(0xa)
  35  839        1        4 A0Power(0xb)
  36  839        1        1 A0Power(0xc)
  37  864        1        1 UartEnabled
  38  868        1        1 A0(0x33e)
  39  670        1        1 UpdateState(A0)
  40  534        1    16649 Status { ier: 0x0, ifr: 0x0, amd_status: 0xc, amd_a0: 0x4 }
  41  564        1        1 NICPowerEnableLow(false)
  42  670        1        1 UpdateState(A0PlusHP)
  43  534        1      257 Status { ier: 0x0, ifr: 0x0, amd_status: 0xc, amd_a0: 0x4 }
  44  683        1        1 SetState { prev: A0PlusHP, next: A2, why: HostBootFailure, now: 0xb580a }
  45 1422        1        1 V3P3SysA0VOut(Volts(3.3105469))
  46  692        1        1 PGStatus { b_pg: 0xff, c_pg: 0x3, nic: 0x7f }
  47  698        1        1 SMStatus { a1: 0x5, a0: 0xc }
  48  703        1        1 PowerControl(0x3)
  49  707        1        1 InterruptFlags(0x0)
  50  670        1        1 UpdateState(A2)
  51 1422        1        1 V3P3SysA0VOut(Volts(3.2480469))
  52  920        1        1 A2
  53 1422        1        1 V3P3SysA0VOut(Volts(1.1953125))
  54 1422        1        1 V3P3SysA0VOut(Volts(0.9921875))
  55  534        1        1 Status { ier: 0x0, ifr: 0x0, amd_status: 0xb, amd_a0: 0xb }
  56  683        1        1 SetState { prev: A2, next: A0, why: HostBootFailure, now: 0xb6c66 }
  57 1422        1        1 V3P3SysA0VOut(Volts(0.09765625))
  58  692        1        1 PGStatus { b_pg: 0x0, c_pg: 0x0, nic: 0x0 }
  59  698        1        1 SMStatus { a1: 0x0, a0: 0x0 }
  60  703        1        1 PowerControl(0x0)
  61  707        1        1 InterruptFlags(0x0)
  62  749        1        1 A1Status(0x2)
  63  749        1      100 A1Status(0x3)
  64  749        1        1 A1Status(0x5)
  65  767        1        1 CPUPresent(true)
  66  777        1        1 Coretype { coretype: true, sp3r1: true, sp3r2: false }
  67  805        1       11 A0Status(0x1)
  68  805        1      306 A0Status(0x2)
  69  805        1        1 A0Status(0x4)
  70  805        1        2 A0Status(0x6)
  71  805        1        1 A0Status(0x7)
  72  827        1        1 RailsOn
  73  839        1        1 A0Power(0x7)
  74  839        1        9 A0Power(0xa)
  75  839        1        4 A0Power(0xb)
  76  839        1        1 A0Power(0xc)
  77  864        1        1 UartEnabled
  78  868        1        1 A0(0x36a)
  79  670        1        1 UpdateState(A0)
  80  534        1      291 Status { ier: 0x0, ifr: 0x0, amd_status: 0xc, amd_a0: 0x4 }
```

</details>